### PR TITLE
feat: support for different MetadataStores in VectorStore

### PR DESCRIPTION
**Pull request by konrad-czarnota-ds**

### CODE REVIEW

File: packages/ragbits-core/src/ragbits/core/vector_store/chromadb_store.py [hunk 1]
```diff
@@ -10,20 +10,29 @@
...
 from ragbits.core.embeddings import Embeddings
+from ragbits.core.metadata_store.base import MetadataStore
 from ragbits.core.utils.config_handling import get_cls_from_config
 from ragbits.core.vector_store import VectorDBEntry, VectorStore
 
 
 class ChromaDBStore(VectorStore):
     """Class that stores text embeddings using [Chroma](https://docs.trychroma.com/)"""
 
+    CHROMA_IDS_KEY = "ids"
```
Comment (mhordynski) [inline packages/ragbits-core/src/ragbits/core/vector_store/chromadb_store.py:21]: tbh I'm not a fan of extracting all constants. It makes the code harder to read later on. Additionally, it clutters the ChromaDBStore public interface; users of this class will receive autosuggestions with a lot of unnecessary variables after typing `ChromaDBStore.`

File: packages/ragbits-core/src/ragbits/core/vector_stores/chroma.py [hunk 2]
```diff
@@ -80,14 +104,20 @@ async def store(self, entries: list[VectorStoreEntry]) -> None:
...
-        ]
-        self._collection.add(ids=ids, embeddings=embeddings, metadatas=metadatas)  # type: ignore
+
+        if self.metadata_store is not None:
+            for key, meta in zip(ids, [entry.metadata for entry in entries], strict=False):
+                await self.metadata_store.store(key, meta)
+            metadata_to_store = None
+        else:
+            metadata_to_store = [
+                {self.METADATA_INNER_KEY: json.dumps(entry.metadata, default=str)} for entry in entries
```
Comment (micpst) [inline packages/ragbits-core/src/ragbits/core/vector_stores/chroma.py:114]: I think we should put the raw entry metadata here, serializing it to a string makes it useless for per-field queries

File: packages/ragbits-core/src/ragbits/core/metadata_store/base.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    async def get_all(self) -> dict:
+        """
+        Returns all keys with associated metadata
+
+        Returns:
+            metadata dict for all entries
+        """
+
+    @abc.abstractmethod
+    async def store_global(self, metadata: dict) -> None:
```
Comment (micpst) [inline packages/ragbits-core/src/ragbits/core/metadata_store/base.py:56]: I don't think we need this for now.

File: packages/ragbits-core/src/ragbits/core/metadata_store/base.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        """
+        Store key value pairs of metadata that is shared across entries
+
+        Args:
+            metadata: common key value pairs for the whole collection
+        """
+
+    @abc.abstractmethod
+    async def get_global(self) -> dict:
```
Comment (micpst) [inline packages/ragbits-core/src/ragbits/core/metadata_store/base.py:65]: ditto

File: packages/ragbits-core/src/ragbits/core/metadata_store/base.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+        Args:
+            key: key to use
+
+        Returns:
+            metadata dict associated with a given key
+        """
+
+    @abc.abstractmethod
+    async def get_all(self) -> dict:
```
[the remark this answers is not in the document]
Comment (konrad-czarnota-ds) [reply packages/ragbits-core/src/ragbits/core/metadata_store/base.py]: deleted

File: packages/ragbits-core/src/ragbits/core/vector_stores/chroma.py [hunk 6]
```diff
@@ -32,24 +44,36 @@ def __init__(
...
             The collection.
         """
-        return self._client.get_or_create_collection(
+        if self._collection is not None:
+            return self._collection
+
+        global_metadata = {"hnsw:space": self._distance_method}
+        if self.metadata_store is not None:
+            await self.metadata_store.store_global(global_metadata)
+            metadata_to_store = None
```
[the remark this answers is not in the document]
Comment (konrad-czarnota-ds) [reply packages/ragbits-core/src/ragbits/core/vector_stores/chroma.py]: fixed, thanks for pointing this out!

File: packages/ragbits-core/src/ragbits/core/vector_stores/chroma.py (hunk 2 above)
Comment (konrad-czarnota-ds) [reply packages/ragbits-core/src/ragbits/core/vector_stores/chroma.py:107]: Created a task for it: https://github.com/deepsense-ai/ragbits/issues/157

File: packages/ragbits-core/src/ragbits/core/metadata_store/base.py (hunk 3 above)
Comment (konrad-czarnota-ds) [reply packages/ragbits-core/src/ragbits/core/metadata_store/base.py]: Deleted

File: packages/ragbits-core/src/ragbits/core/metadata_store/base.py (hunk 4 above)
Comment (konrad-czarnota-ds) [reply packages/ragbits-core/src/ragbits/core/metadata_store/base.py]: Deleted

File: packages/ragbits-core/src/ragbits/core/vector_stores/base.py [hunk 7]
```diff
@@ -29,9 +31,12 @@ class VectorStore(ABC):
     A class with an implementation of Vector Store, allowing to store and retrieve vectors by similarity function.
     """
 
-    def __init__(self, default_options: VectorStoreOptions | None = None) -> None:
+    metadata_store: MetadataStore | None
```
Comment (micpst) [inline packages/ragbits-core/src/ragbits/core/vector_stores/base.py:34]: This'll be buggy, if we create multiple instances of `VectorStore`. Let's make it a private attribute instance.

File: packages/ragbits-core/src/ragbits/core/vector_stores/chroma.py [hunk 8]
```diff
@@ -32,24 +44,29 @@ def __init__(
...
             default_options: The default options for querying the vector store.
+            metadata_store: The metadata store to use.
         """
-        super().__init__(default_options)
+        super().__init__(default_options, metadata_store)
         self._client = client
         self._index_name = index_name
         self._distance_method = distance_method
-        self._collection = self._get_chroma_collection()
+        self._collection: Collection | None = None
```
Comment (micpst) [inline packages/ragbits-core/src/ragbits/core/vector_stores/chroma.py:53]: Why change here?

File: packages/ragbits-core/src/ragbits/core/vector_stores/chroma.py [hunk 9]
```diff
@@ -32,24 +44,29 @@ def __init__(
...
 
-    def _get_chroma_collection(self) -> Collection:
+    async def _get_chroma_collection(self) -> Collection:
```
Comment (micpst) [inline packages/ragbits-core/src/ragbits/core/vector_stores/chroma.py:55]: No need to be async

File: packages/ragbits-document-search/src/ragbits/document_search/ingestion/providers/unstructured/images.py [hunk 10]
```diff
@@ -95,12 +98,18 @@ async def _to_image_element(
 
         img_bytes = crop_and_convert_to_bytes(image, top_x, top_y, bottom_x, bottom_y)
         prompt = _ImagePrompt(_ImagePromptInput(images=[img_bytes]))
-        if self._llm:
-            image_description = await self._llm.generate(prompt=prompt)
-        elif self.image_describer:
-            image_description = await self.image_describer.get_image_description(prompt=prompt)
-        else:
-            image_description = ""
+        if self.image_describer is None:
```
Comment (micpst) [inline packages/ragbits-document-search/src/ragbits/document_search/ingestion/providers/unstructured/images.py:101]: I think this shouldn't be affected by this PR.

File: packages/ragbits-core/src/ragbits/core/vector_stores/chroma.py (hunk 8 above)
Comment (konrad-czarnota-ds) [reply packages/ragbits-core/src/ragbits/core/vector_stores/chroma.py]: When I had global metastore with async methods I needed to do a lazy loading, should have deleted it with previous commit, thanks for pointing it out.

File: packages/ragbits-core/src/ragbits/core/vector_stores/chroma.py (hunk 9 above)
Comment (konrad-czarnota-ds) [reply packages/ragbits-core/src/ragbits/core/vector_stores/chroma.py]: Same as above, changed

**Review by micpst — APPROVED** [review]

**Pull request merged**